src/main/java/de/florianmichael/viafabricplus/injection/mixin/fixes/minecraft/screen/hud/MixinPlayerListHud.java

```diff
@@ -0,0 +1,104 @@
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Mixin(PlayerListHud.class)
+public abstract class MixinPlayerListHud implements IPlayerListHud {
+
+    @Unique
+    private static final Comparator<PlayerListEntry> viaFabricPlus$FIFO_COMPARATOR
+            = Comparator.comparingInt(ple -> ((IPlayerListEntry) ple).viaFabricPlus$getIndex());
+
+    @Shadow
+    @Final
+    private MinecraftClient client;
+
+    @Unique
+    private int viaFabricPlus$maxSlots = 60;
+    @Unique
+    private boolean viaFabricPlus$hideSkins = true;
+
+    @Inject(method = "collectPlayerEntries", at = @At("HEAD"), cancellable = true)
+    private void collectPlayerEntries(CallbackInfoReturnable<List<PlayerListEntry>> result) {
+        if (VisualSettings.global().enableLegacyTablist.isEnabled()) {
+            result.setReturnValue(this.client.player.networkHandler.getListedPlayerListEntries().stream()
+                    .sorted(viaFabricPlus$FIFO_COMPARATOR)
+                    .limit(viaFabricPlus$maxSlots)
+                    .collect(Collectors.collectingAndThen(Collectors.toList(), this::viaFabricPlus$transpose)));
+        } else {
+            viaFabricPlus$hideSkins = false;
+        }
+    }
+
+    @Redirect(method = "render", at = @At(value = "INVOKE", target = "Lnet/minecraft/network/ClientConnection;isEncrypted()Z"))
+    private boolean hideSkins(ClientConnection instance) {
+        return !viaFabricPlus$hideSkins && instance.isEncrypted();
+    }
+
+    @Unique
+    private List<PlayerListEntry> viaFabricPlus$transpose(final List<PlayerListEntry> list) {
+        // Only bother transposing if we know the list is full
+        if (list.size() != viaFabricPlus$maxSlots) {
+            viaFabricPlus$hideSkins = list.stream().noneMatch(e -> e.getProfile().getProperties().containsKey("textures"));
+            return list;
+        }
+
+        final List<PlayerListEntry> result = new ArrayList<>(list.size());
+
+        final int columns = viaFabricPlus$maxSlots / PlayerListHud.MAX_ROWS;
+        boolean anyHasSkinData = false;
+        for (int i = 0; i < viaFabricPlus$maxSlots; i++) {
+            final int row = i % PlayerListHud.MAX_ROWS;
+            final int col = i / PlayerListHud.MAX_ROWS;
+            final PlayerListEntry current = list.get(row * columns + col);
+            result.add(current);
+            anyHasSkinData = anyHasSkinData || current.getProfile().getProperties().containsKey("textures");
+        }
+        viaFabricPlus$hideSkins = !anyHasSkinData;
+        return result;
+    }
+
+    @Override
+    public void viaFabricPlus$setMaxPlayers(final int maxPlayers) {
+        this.viaFabricPlus$maxSlots = Math.min(200, Math.max(20,
+                ((maxPlayers + PlayerListHud.MAX_ROWS - 1) / PlayerListHud.MAX_ROWS) * PlayerListHud.MAX_ROWS));
```

Comment:
What's the reason for doing **/ PlayerListHud.MAX_ROWS) * PlayerListHud.MAX_ROWS**. Looking at the code it seems to just validate the range of maxPlayers

Reply:
max players can be any integer, sent by the server
the max slots is what is locally used to decide how many columns is the "fully filled" tablist, and is a multiple of 20 (20, 40, 60, 80, 100 and up to 200). This essentially "rounds to the nearest 20"

Reply:
Sounds good, looking at the 1.7 code I can find the same code/logic in there, thank you for confirming.